src/crypto/x448.rs

```diff
@@ -0,0 +1,220 @@
+
+use crate::crypto::{aes_kw, Decryptor, KeyParams};
+use crate::errors::Result;
+use crate::types::{PlainSecretParams, PublicParams};
+
+/// Secret key for X448
+#[derive(Clone, derive_more::Debug, PartialEq, Eq, Zeroize, ZeroizeOnDrop)]
+pub struct SecretKey {
+    #[debug("..")]
+    pub(crate) secret: [u8; 56],
+}
+
+impl KeyParams for SecretKey {
+    type KeyParams = ();
+
+    fn key_params(&self) {}
+}
+
+pub struct EncryptionFields<'a> {
+    /// Ephemeral X448 public key (56 bytes)
+    pub ephemeral_public_point: [u8; 56],
+
+    /// Recipient public key (56 bytes)
+    pub recipient_public: [u8; 56],
+
+    /// Encrypted and wrapped session key
+    pub encrypted_session_key: &'a [u8],
+}
+
+impl Decryptor for SecretKey {
+    type EncryptionFields<'a> = EncryptionFields<'a>;
+
+    fn decrypt(&self, data: Self::EncryptionFields<'_>) -> Result<Vec<u8>> {
+        debug!("X448 decrypt");
+
+        let shared_secret = {
+            // create montgomery point
+            let their_public =
+                x448::PublicKey::from_bytes(&data.ephemeral_public_point).expect("56");
+
+            // private key of the recipient.
+            let our_secret = x448::Secret::from(self.secret);
+
+            // derive shared secret
+            let shared_secret = our_secret.as_diffie_hellman(&their_public).expect("FIXME");
+
+            *shared_secret.as_bytes()
+        };
+
+        // obtain the session key from the shared secret
+        derive_session_key(
+            data.ephemeral_public_point,
+            data.recipient_public,
+            shared_secret,
+            data.encrypted_session_key,
+        )
+    }
+}
+
+/// Obtain the decrypted OpenPGP session key
+///
+/// This helper function performs the steps described in
+/// https://www.rfc-editor.org/rfc/rfc9580.html#name-algorithm-specific-fields-for-x
+pub fn derive_session_key(
+    ephemeral: [u8; 56],
+    recipient_public: [u8; 56],
+    shared_secret: [u8; 56],
+    encrypted_session_key: &[u8],
+) -> Result<Vec<u8>> {
+    let okm = hkdf(&ephemeral, &recipient_public, &shared_secret)?;
+
+    let decrypted_key = aes_kw::unwrap(&okm, encrypted_session_key)?;
+    ensure!(!decrypted_key.is_empty(), "empty key is not valid");
+
+    Ok(decrypted_key)
+}
+
+/// Generate an X448 KeyPair.
+pub fn generate_key<R: Rng + CryptoRng>(mut rng: R) -> (PublicParams, PlainSecretParams) {
+    let mut secret_key_bytes = Zeroizing::new([0u8; 56]);
+    rng.fill_bytes(&mut *secret_key_bytes);
+
+    let secret = x448::Secret::from(*secret_key_bytes); // does clamping
+    let public = *x448::PublicKey::from(&secret).as_bytes();
+
+    (
+        PublicParams::X448 { public },
+        PlainSecretParams::X448(*secret.as_bytes()),
+    )
+}
+
+/// HKDF for X448
+/// https://www.rfc-editor.org/rfc/rfc9580.html#name-algorithm-specific-fields-for-x
+pub fn hkdf(
+    ephemeral: &[u8; 56],
+    recipient_public: &[u8; 56],
+    shared_secret: &[u8; 56],
+) -> Result<[u8; 32]> {
```

Comment:
could this be shared with the hkdf in x25519 by passing in the info and sizes as generics/arguments?

Reply:
I think it can. But the shared function would also need to take a `Digest` as parameter (`Sha256` vs. `Sha512`).

I'm sure this is possible, but I don't know how. And I wonder about the const-generics complexity vs. DRY-ness tradeoff.

Reply:
it's fine for now, maybe just add a todo to explore this for now